### PR TITLE
Support for glowing 3D line overlays

### PR DESCRIPTION
**Pull request by jherico**

### CODE REVIEW

**Comment by jherico** [comment]
## Testing

A test script which exercises a number of overlay types including the new glow line is available [here](http://s3.amazonaws.com/DreamingContent/scripts/tests/overlays.js)

**Comment by jherico** [comment]
<a href="https:&#x2F;&#x2F;trello.com&#x2F;c&#x2F;jbnwDhHH&#x2F;56-make-seek-and-select-lasers-look-more-real"> Make seek and select lasers look more real</a>

**Comment by hifi-gustavo** [comment]
Build console output for commit: fddda925a55b51899049307b53de553caf94895d available at: https://gist.github.com/bb3965d2798412b4ab93e7e665d3d773<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8194.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8194.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 894e11a370a4091dc69273334f1787e3ce8ba7d3 available at: https://gist.github.com/bb3965d2798412b4ab93e7e665d3d773<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8194.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8194.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 703c7a8dd055c520918441a2748f7d974180f3af available at: https://gist.github.com/bb3965d2798412b4ab93e7e665d3d773<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8194.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8194.dmg

**Comment by imgntn** [comment]
some issues w transparency on nvidia that we discussed
capture3

File: libraries/render-utils/src/GeometryCache.cpp [hunk 1]
```diff
@@ -1227,41 +1230,39 @@ void GeometryCache::renderDashedLine(gpu::Batch& batch, const glm::vec3& start,
...
 GeometryCache::BatchItemDetails::BatchItemDetails() :
-    verticesBuffer(NULL),
-    colorBuffer(NULL),
-    streamFormat(NULL),
-    stream(NULL),
-    vertices(0),
-    vertexSize(0),
-    isCreated(false)
-{
+verticesBuffer(NULL),
```
Comment (samcake) [inline libraries/render-utils/src/GeometryCache.cpp]: Why?

File: libraries/render-utils/src/glowLine.slg [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+layout(location = 0) in vec4 inColor[];
+
+layout(location = 0) out vec4 outColor;
+layout(location = 1) out vec3 outLineDistance;
+
+layout(lines) in;
+layout(triangle_strip, max_vertices = 24) out;
+
+struct TransformCamera {
+    mat4 _view;
```
Comment (samcake) [inline libraries/render-utils/src/glowLine.slg]: SO here, please Include Transform.slh which will provide the current master representation of the TransformCamera etc

File: libraries/render-utils/src/GeometryCache.cpp (hunk 1 above)
Comment (jherico) [inline libraries/render-utils/src/GeometryCache.cpp]: I accidentally hit the 'format document' keyboard shortcut when I meant to hit 'comment block' at one point.  I didn't realize it until after I'd made the PR.  If you view the PR diff with whitespace changes turned off, you'll see that it's basically unchanged except for the new stuff for the glow line effect.

File: libraries/render-utils/src/glowLine.slg (hunk 2 above)
Comment (jherico) [inline libraries/render-utils/src/glowLine.slg]: Ah, sorry, I meant to do that..  this is a holdover from when I was doing testing by loading a pre-scribed version of the file from disk.  Will fix.

**Comment by hifi-gustavo** [comment]
Build console output for commit: c644c87e6945a4e13da0b8a189935e06fc4b11d3 available at: https://gist.github.com/bb3965d2798412b4ab93e7e665d3d773<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8194.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8194.dmg

File: libraries/render-utils/src/GeometryCache.cpp [hunk 3]
```diff
@@ -1430,13 +1432,103 @@ void GeometryCache::renderLine(gpu::Batch& batch, const glm::vec2& p1, const glm
...
+    // Compile the shaders
+    static std::once_flag once;
+    std::call_once(once, [&] {
+        auto state = std::make_shared<gpu::State>();
+        auto VS = gpu::Shader::createVertex(std::string(glowLine_vert));
+        auto GS = gpu::Shader::createGeometry(std::string(glowLine_geom));
+        auto PS = gpu::Shader::createPixel(std::string(glowLine_frag));
+        auto program = gpu::Shader::createProgram(VS, GS, PS);
+        state->setCullMode(gpu::State::CULL_NONE);
+        state->setDepthTest(true, true, gpu::LESS_EQUAL);
```
Comment (samcake) [inline libraries/render-utils/src/GeometryCache.cpp]: Try without Writing in the Z buffer

**Comment by imgntn** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: ae1dd29a7a012414742ed9142ae17715b3291810 available at: https://gist.github.com/bb3965d2798412b4ab93e7e665d3d773<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8194.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8194.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: ae1dd29a7a012414742ed9142ae17715b3291810 available at: https://gist.github.com/bb3965d2798412b4ab93e7e665d3d773<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8194.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8194.dmg

**Pull request merged**